{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "**Anlysis of the colonization data **  "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "\n",
      "\n",
      "It is my thought to keep this notebook as evidence of my analysis of the large quantity of germfree data the lab has produced.  I have had several analyses of parts of this data set, but this the complete set including some reruns of the samples that produced few reads in the past.  The goal here is to identify whether or not deterministic factors are at play in the colonization of the gut by bacterial communities.  Once the data has been properly processed by the mothur pipeline we will aim to identify trends that arise in the colonization of germ free (gf) mice by mouse communities.  We will then observe whether or not these trends hold when human fecal communities are used as innocula, and the ultimate test of the hypothesis will be whether or not various trends are maintained when soil, which contains a diverse, but not gut adapted, community is used as innoculum."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": []
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Initial anlysis"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Before begining the mothur pipline it was helpful to create a .files file containing sample names that were of a unified structure.  This data set contains many different experiments ran by different scientists and they each have their own naming skeme, some more helpful than others.  The needed file for the make.contigs command was created below and named detrm.files because the goal is to show the colonization process is deterministic (aka detrm). "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/usr/bin/python\n",
      "\n",
      "import sys\n",
      "import re\n",
      "\n",
      "#file = open(sys.argv[1],'r')\n",
      "#outfile = open(sys.argv[2],'w')\n",
      "file = open('fastq.txt','r')\n",
      "outfile = open('colonus.files','w')\n",
      "\n",
      "# ------------naming skemes for innocula--------------------------------\n",
      "\n",
      "alyx_contam_inocA = re.compile('(\\d+)A-inoculum') # 268A-inoculum_S64_L001_R1_001.fastq\n",
      "niel_inoc = re.compile('([CH]\\D{5,6}\\d+)-\\d{4}')  # Cancer16-2619_S76_L001_R1_001.fastq or Healthy....\n",
      "alyx_contam_inocB = re.compile('DA(\\d+)')     # DA6_S380_L001_R1_001.fastq\n",
      "matt_conv2 = re.compile('ConvIDneg(\\d+)')    # ConvIDneg20_S137_L001_R1_001.fastq\n",
      "matt_syn2_inoc = re.compile('Mix(\\d)IDneg(\\d+)')  # Mix1IDneg20_S138_L001_R1_001.fastq\n",
      "\n",
      "# -------------naming skemes for scientists and experiments --------------------\n",
      "matt_soil = re.compile('(\\d{4})-.*-(\\d+)')           #1810-D-10_S308_L001_R1_001.fastq\n",
      "matt_soilNT= re.compile('(NT)-.*-(\\d+)')              #N-T-dn-10_S282_L001_R1_001.fastq  The N-T is adjusted below to 'NT' \n",
      "matt_syn1 = re.compile('([A-C]\\d)-(\\d{2})-(\\d+)')      #C4-11-5_S140_L001_R1_001.fastq\n",
      "matt_syn2 = re.compile('([A-C])(\\d{3})D.*?(\\d{1,2})')     #C901D0_S238_L001_R1_001.fastq\n",
      "niel_2X = re.compile('([a-b]\\d)-(.*)-.*')               # b4-nt-d00_S155_L001_R1_001.fastq\n",
      "niel_4XJT = re.compile('([1-4][A-D])-(.*)-(\\d+)')         #1A-nt-5_S329_L001_R1_001.fastq\n",
      "niel_no_= re.compile('([a-d])([^-]*)dn?(\\d+)')            #Niel's samples that are a622d0_S219_L001_R1_001.fastq\n",
      "alyx_contam = re.compile('(\\d+[AB])-(.*)-(D[-\\d]\\d*)')      #268B-NT-D-7_S96_L001_R1_001.fastq\n",
      "alyx_treat = re.compile('([NYC]\\D*)-(\\w+)-*D-?(\\d+)')       # NCHL-106-D0_S243_L001_R1_001.fastq Also Careyblair treatments\n",
      "\n",
      "\n",
      "#----------control----------------\n",
      "control_matt = re.compile('C(\\d)-5-(\\d+)')\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "# -------------Set up lists---------------------------\n",
      "r1 = []   # fastq1\n",
      "r2 = []   #fastq2\n",
      "samples = []  #Sample name\n",
      "missing =[]    # those that miss the boat on the skemes above\n",
      "r = 1\n",
      "\n",
      "\n",
      "#------------Loop through the fastq.txt file\n",
      "for line in file:\n",
      "    line = line.strip()           #remove the whitespace\n",
      "    if r%2!=0:                    #odd\n",
      "        fastq1=line               # Save the *.fastq name for the second colum of the files file.\n",
      "        line = line.split('_')     # splits fastq file name at the sample name and saves this as name\n",
      "        name1 = line[0]            # Save the original name to be matched to the other fastq later\n",
      "        name = name1.replace('N-T','NT') # replace the N-T in matt's with NT for mouse\n",
      "        \n",
      "        if name == '846dn03':  # corrects a typo which labeled one of niels samples incorrectly\n",
      "            name = 'd846dn03'\n",
      "        if name == 'C03Dneg18':   # corrects a typo which labeled one of matt's syn2 samples incorrectly\n",
      "            name = 'C903Dneg18'    \n",
      "# -------------Determining how to rename the samples-------------------------------------------------\n",
      "\n",
      "## INOCULA ###\n",
      "        alyx_con_inocA = alyx_contam_inocA.match(name)\n",
      "        n_inoc = niel_inoc.match(name)   \n",
      "        alyx_con_inocB = alyx_contam_inocB.match(name)\n",
      "        syn2_conv = matt_conv2.match(name)\n",
      "        syn2_mix = matt_syn2_inoc.match(name)\n",
      "##Control##\n",
      "        control = control_matt.match(name)\n",
      "##Samples ##       \n",
      "        soil = matt_soil.match(name)\n",
      "        soilNT = matt_soilNT.match(name)\n",
      "        syn1 = matt_syn1.match(name)\n",
      "        syn2 = matt_syn2.match(name)\n",
      "        a1b4=niel_2X.match(name)\n",
      "        A14D=niel_4XJT.match(name)\n",
      "        no_ = niel_no_.match(name) \n",
      "        contam = alyx_contam.match(name)\n",
      "        treat = alyx_treat.match(name)\n",
      "        \n",
      "## INOCULA ###        \n",
      "        if alyx_con_inocA:\n",
      "            scientist='alyx'\n",
      "            group = alyx_con_inocA.group(1)  # number\n",
      "            cage = 'A'\n",
      "            mouse= 'INOC'\n",
      "            day = mouse\n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test = name1     # save original name to be matched with other fastq later\n",
      "        \n",
      "        elif alyx_con_inocB:\n",
      "            scientist='alyx'\n",
      "            group = alyx_con_inocB.group(1) # number \n",
      "            cage = 'B'\n",
      "            mouse= 'INOC'\n",
      "            day = mouse\n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test = name1  \n",
      "        elif n_inoc:\n",
      "            scientist = 'niel'\n",
      "            if n_inoc.group(1)=='Cancer7':\n",
      "                group = 'C1'\n",
      "            elif n_inoc.group(1)=='Cancer16':  \n",
      "                group = 'C2'\n",
      "            elif n_inoc.group(1)=='Cancer21':              \n",
      "                group = 'C3'\n",
      "            elif n_inoc.group(1)=='Healthy2':    \n",
      "                group = 'H1'\n",
      "            elif n_inoc.group(1)=='Healthy11':\n",
      "                group = 'H2'\n",
      "            elif n_inoc.group(1)=='Healthy18':    \n",
      "                group = 'H3'\n",
      "            else:\n",
      "            \tgroup = 'WRONG'    # for decoding\n",
      "            cage = 'INOC'    \n",
      "            mouse = cage\n",
      "            day = cage    \n",
      "                \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test = name1\n",
      "        \n",
      "        elif syn2_conv:   # conventialized mice for control in syn2 experiment \n",
      "            scientist='matt'\n",
      "            group = 'syn2'\n",
      "            cage = 'Conv'\n",
      "            mouse = 'INOC'\n",
      "            day = str(20-int(syn2_conv.group(1))) # Dneg 19 = Day 1 There were three days of inocculation Dneg 22-20\n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test = name1\n",
      "        elif syn2_mix:\n",
      "            scientist ='matt'\n",
      "            group = 'syn2'\n",
      "            if syn2_mix.group(1)=='1':\n",
      "                cage = 'A'\n",
      "            elif syn2_mix.group(1)=='2':\n",
      "                cage = 'B'\n",
      "            else:\n",
      "                cage = 'WRONG'\n",
      "            \n",
      "            mouse = 'INOC'\n",
      "            day = str(20-int(syn2_mix.group(2))) # Dneg 19 = Day 1 There were three days of inocculation Dneg 22-20\n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test = name1     \n",
      "                \n",
      "                  \n",
      "        elif name == 'soil-Inoc':\n",
      "            scientist = 'matt'\n",
      "            group = 'soil'\n",
      "            cage = 'INOC'\n",
      "            mouse = cage\n",
      "            day = cage\n",
      "            \n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test = name1  \n",
      "            \n",
      "        elif name == 'Cult-inoc-10-24':\n",
      "            scientist = 'matt'\n",
      "            group = 'syn1'\n",
      "            cage = 'syn'\n",
      "            mouse = 'INOC'\n",
      "            day = mouse\n",
      "            \n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test = name1 \n",
      "       \n",
      "        elif name == 'Feces-inoc-10-24':\n",
      "             scientist = 'matt'\n",
      "             group = 'syn1'\n",
      "             cage = 'Conv'\n",
      "             mouse = 'INOC'\n",
      "             day = mouse\n",
      "            \n",
      "            \n",
      "             newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "             samples.append(newname)\n",
      "             r1.append(fastq1)\n",
      "             test = name1      \n",
      "##Control##\n",
      "        elif control:\n",
      "            scientist = 'matt'\n",
      "            group = 'Control'\n",
      "            if int(control.group(1)) <3:  # mice 1 and 2 were cohoused. 3-5 were also cohoused \n",
      "                cage = 'A'\n",
      "            else:\n",
      "                cage = 'B'\n",
      "            mouse = control.group(1)\n",
      "            day = control.group(2)  # Experiment began on May 1 so we can just take the day\n",
      "\n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test = name1\n",
      "\n",
      "## Samples ##                \n",
      "                \n",
      "         # Matt soil\n",
      "        elif soil:\n",
      "            scientist='matt'\n",
      "            group = 'soil'\n",
      "            cage=soil.group(1)\n",
      "            mouse = soil.group(1)\n",
      "            day = str(26-int(soil.group(2)))\n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test = name1\n",
      "        elif soilNT: \n",
      "            scientist='matt'\n",
      "            group = 'soil'\n",
      "            cage=soilNT.group(1)\n",
      "            mouse = soilNT.group(1)\n",
      "            day = str(26-int(soilNT.group(2)))\n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "       \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test=name1\n",
      "        \n",
      "        # Matt Synthetic\n",
      "        elif syn1:\n",
      "            scientist='matt'\n",
      "            group = 'syn1'\n",
      "            cage=syn1.group(1)[0]   # All in format 'A2' were A is the cage and 2 is the mouse\n",
      "            mouse = syn1.group(1)[1]\n",
      "            if int(syn1.group(2)) == 10:         # The days were in date format with 10-23 as day -1 with 10-25 as day 1 \n",
      "                    day = str(int(syn1.group(3)) - 24)\n",
      "            if int(syn1.group(2)) == 11:\n",
      "                    day = str(int(syn1.group(3)) + 6) # 11-1 was really day 7\n",
      "            \n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test=name1\n",
      "            \n",
      "        elif syn2:\n",
      "            #print('Matched a syn2!')\n",
      "            scientist = 'matt'\n",
      "            group = 'syn2'\n",
      "            cage= syn2.group(1)\n",
      "            mouse = syn2.group(2)\n",
      "            day = str(20-int(syn2.group(3))) # there were 3 days of Inoculation they are -2,-1, and 0 day 1 corresponds to Dneg19    \n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            \n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test=name1\n",
      "            \n",
      "        # Niel a1a2 b3b4  Used reg exp to identify but then for some reason I split the name variable to allocate, It was just easier\n",
      "        elif a1b4:\n",
      "            name=name.split('-')    \n",
      "            if name[0]=='a1' or name[0]=='a2':  # Had same innocula\n",
      "                group = 'H3'\n",
      "            elif name[0]=='b3' or name[0]=='b4':\n",
      "                group='C1'\n",
      "            else:\n",
      "                group= 'WRONG'        \n",
      "            scientist='niel'\n",
      "            cage = name[0]\n",
      "            mouse = name[1]\n",
      "            if 'd' in name[-1]:           # Some had d in the day section\n",
      "                try: \n",
      "                    m = re.search('(?<=\\w)\\d+',name[-1])  # take the digits after the last letter character\n",
      "                    \n",
      "                    day = str(22-int(m.group(0))) # d0 becomes day 22\n",
      "                except ValueError:\n",
      "                    day = 'WRONG'\n",
      "            else:\n",
      "                try:\n",
      "                    day = str(22-int(name[-1]))\n",
      "                except ValueError:\n",
      "                    day = 'WRoNG'\n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test=name1\n",
      "            \n",
      "            #Niel 1A-4D\n",
      "        elif A14D:\n",
      "            if A14D.group(1) =='1A':\n",
      "                group= 'H1'\n",
      "            elif A14D.group(1) =='2B':\n",
      "                group = 'H2'\n",
      "            elif A14D.group(1)=='3C':\n",
      "                group = 'C2'\n",
      "            elif A14D.group(1) == '4D':\n",
      "                group = 'C3'            \n",
      "            else:\n",
      "                group = 'WRONG'\n",
      "            scientist = 'niel'\n",
      "            cage = A14D.group(1)\n",
      "            mouse = A14D.group(2)\n",
      "            day = str(22-int(A14D.group(3)))    # again there were 22 days starting at day 0\n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test=name1\n",
      "        #Neils tricky samples \n",
      "        elif no_:\n",
      "        \n",
      "            if no_.group(1) =='a':\n",
      "                group= 'H1'\n",
      "            elif no_.group(1) =='b':\n",
      "                group = 'H2'\n",
      "            elif no_.group(1)=='c':\n",
      "                group = 'C2'\n",
      "            elif no_.group(1) == 'd':\n",
      "                group = 'C3'            \n",
      "            else:\n",
      "                group = 'WRONG'\n",
      "\n",
      "            if no_.group(1) =='a':\n",
      "                cage= '1A'\n",
      "            elif no_.group(1) =='b':\n",
      "                cage = '2B'\n",
      "            elif no_.group(1)=='c':\n",
      "                cage = '3C'\n",
      "            elif no_.group(1) == 'd':\n",
      "                cage = '4D'            \n",
      "            else:\n",
      "                cage = 'WRONG'\n",
      "\n",
      "            scientist = 'niel'\n",
      "            mouse = no_.group(2)\n",
      "            day = str(22-int(no_.group(3)))    # again there were 22 days starting at day 0\n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test=name1\n",
      "\n",
      "        #Alyx contaminated\n",
      "        elif contam:\n",
      "            scientist = 'alyx'\n",
      "            group = contam.group(1)[0:-1] # take cage number but save the letter for the cage cattegory in the next line\n",
      "            cage = contam.group(1)[-1]\n",
      "            mouse = contam.group(2)\n",
      "            day = contam.group(3)\n",
      "            if '-' in day:    # Negative days were denoted D_#\n",
      "                day = day.split('-')\n",
      "                day = day[-1]\n",
      "                day = str(15-int(day))   # 14 days starting at day 0\n",
      "            elif 'D0' in day:\n",
      "                day = '15'\n",
      "            else:\n",
      "                day = day.split('D') \n",
      "                day = str(15+int(day[-1]))         # these are for the few samples that extended beyond day 0 in the original experiment\n",
      "\n",
      "\n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test=name1\n",
      "           \n",
      "        #Alyx's treated samples (Chloroform and Carey Blair)\n",
      "        elif treat:\n",
      "            scientist = 'alyx'\n",
      "            if 'CHL' in treat.group(1):\n",
      "                group = 'Chloroform'\n",
      "            elif treat.group(1)=='N' or 'Y' or 'NC':\n",
      "                group = 'CairyBlair'\n",
      "            cage = treat.group(1)\n",
      "            mouse = treat.group(2).strip('-D')  # some samples are mouseD- others are mouse-D- this removes the D or the -D from the mouse string        \n",
      "            day = str(15-int(treat.group(3)))\n",
      "            \n",
      "            newname = scientist + '_' + group + '_' + cage + '_' + mouse + '_' + day\n",
      "            samples.append(newname)\n",
      "            r1.append(fastq1)\n",
      "            test=name1\n",
      "   #------------If the name did not match the conditions above add it to the missing list-----------------------------------                   \n",
      "        else:\n",
      "            missing.append(name)    \n",
      "  # ------the other fastq --------even lines-------------      \n",
      "    if r%2==0:  #even\n",
      "        fastq2=line \n",
      "        line = line.split('_')     # splits fastq file name at the sample name and saves this as name\n",
      "        name2 = line[0]\n",
      "        if name2==test:          # test to see if name matches that of the last successful fastq\n",
      "            r2.append(fastq2)\n",
      "    r += 1\n",
      "#--------------Replace 'bad' reads with fastq from rerun samples-----------------------------------------------------\n",
      "file.close\n",
      "\n",
      "file = open('fastq.txt','r')\n",
      "\n",
      "repeat = re.compile('^b-(.*fastq)')  # all the reruns have the form 'b-name_....fastq'\n",
      "\n",
      "r=1\n",
      "r1_replacements = []  # keep track of replacements for the sake of debugging\n",
      "r2_replacements=[]\n",
      "for line in file:\n",
      "    line = line.strip()\n",
      "    fastq=line \n",
      "    rep = repeat.match(fastq)        # go through the original fastq.txt and pull out a b- \n",
      "    if rep:\n",
      "        searchfastq = rep.group(1)  # no b-\n",
      "        searchname = searchfastq.split('_')[0]  # Everything before the _ in the fastq\n",
      "           \n",
      "        # some samples had slightly different names upon resequencing them mostly due to typos\n",
      "        if searchname == '268-B-NT-D0':\n",
      "            searchname = '268B-NT-D0'\n",
      "        elif searchname == '518B-18-D-5':\n",
      "            searchname = '581B-18-D-5'        \n",
      "        elif searchname == '286A-1-D-11':\n",
      "            searchname = '268A-1-D-11'\n",
      "        elif searchname == '581B-2-d1':\n",
      "            searchname = '581B-2-D1'\n",
      "        elif searchname == 'b3-1357-19':\n",
      "            #print('found it')\n",
      "            searchname = 'b4-1357-19'\n",
      "       \n",
      "        elif 'NT' in searchname:\n",
      "            searchname=searchname.replace('NT-dn','N-T-dn')   # just makes sure NT from matt will match the origianl sample names\n",
      "    \n",
      "    \n",
      "    \n",
      "    \n",
      "    \n",
      "        if r%2!=0: #odd\n",
      "    \t    for l in range(len(r1)):         #r1 is from the odd lines and so should only include R1 reads so searching for name is fine\n",
      "                if searchname + '_' in r1[l]:  # find the orginal fastq and replace it\n",
      "                    r1[l] = fastq\n",
      "\n",
      "            \n",
      "        if r%2==0: #even\n",
      "           \n",
      "            for l in range(len(r2)):         #r2 is from the even lines and so should only include R2 reads so searching for name is fine\n",
      "                if searchname +'_' in r2[l]:\n",
      "                   r2[l] = fastq\n",
      "    r+=1    \n",
      "    \n",
      "    \n",
      "##---------------Check for missed samples--------------------------------------------#\n",
      "\n",
      "for j in range(len(r1)):         \n",
      "    replacement = repeat.match(r1[j]) # see if the fastq matches a one that was inserted above during the replacement step\n",
      "    if replacement:\n",
      "        replacement = replacement.group(0).split('_')[0]\n",
      "        r1_replacements.append(replacement)   #Add the name to the replacement list\n",
      "        \n",
      "for i in range(len(r2)):        \n",
      "    replacement = repeat.match(r2[i])\n",
      "    if replacement:\n",
      "        replacement = replacement.group(0).split('_')[0]\n",
      "        r2_replacements.append(replacement)  # Does the same as the loop above but this time for the second fastq column\n",
      "        \n",
      "# Make sure r1 and r2 are the same\n",
      "print('There were replaced differentlty between r1 and r2')\n",
      "               \n",
      "different1 = set(r1_replacements).difference(set(r2_replacements))\n",
      "different2 = set(r2_replacements).difference(set(r1_replacements))\n",
      "different = set(different1).union(set(different2))\n",
      "print(different)    \n",
      "    \n",
      "print('You forgot these you fool!')\n",
      "\n",
      "\n",
      "replacements = r1_replacements # this is good so long as there is no difference between r1 and r2\n",
      "\n",
      "forgot = set(missing).difference(set(replacements))   #Checkes that those that were missing (which are mostly the 'b-...' reruns are now included\n",
      "\n",
      "print(len(forgot))\n",
      "print(forgot)\n",
      "#-----Write to .file file------------------------------------\n",
      "for i in range(len(samples)):\n",
      "    outfile.write(samples[i]+'\\t'+r1[i]+'\\t'+r2[i]+'\\n')\n",
      "    \n",
      "file.close()\n",
      "outfile.close()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "There were replaced differentlty between r1 and r2\n",
        "set([])\n",
        "You forgot these you fool!\n",
        "23\n",
        "set(['I3-D6', 'I3-D7', 'I1-D7', 'I1-D5', 'I1-D8', 'I3-D8', 'I4-D8', 'C2-2-10-25', 'I4-D7', 'I4-D6', 'I2-D7', 'I5-D7', 'I5-D5', 'I5-D8', 'I2-D8', 'C5-colon-cont', 'C1-colon-cont', 'C4-colon-cont', 'I2-D6', 'I6-D8', 'I6-D7', 'I6-D6', 'C2-colon-cont'])\n"
       ]
      }
     ],
     "prompt_number": 37
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Then the following was ran in the make.batch file on axiom using the make.pbs script"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "make.contigs(file=colonus.files, processors=16)\n",
      "summary.seqs(fasta=colonus.trim.contigs.fasta, processors=16)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 29
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for j in range(len(r1)):         #r1 is from the odd lines and so should only include R1 reads so searching for name is fine\n",
      "    replacement = repeat.match(r1[j])\n",
      "    replacement = replacement.group(0).split('_')[0]\n",
      "    r1_replacements.append(replacement)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "NameError",
       "evalue": "name 'replacement' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[0;32m<ipython-input-11-b9b4bf1d402a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mreplacement\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[0;31mNameError\u001b[0m: name 'replacement' is not defined"
       ]
      }
     ],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}